### PR TITLE
Explore TZ-aware date type in Pydantic v2 (GSI 418)

### PR DESCRIPTION
**Pull request by TheByronHimes**
### TL;DR
The point of this exploration was to see if the new `AwareDatetime` in Pydantic v2 is able to replace our DateTimeUTC. 
The answer is... "sorta." I'm not sure if it's worth it to switch since what we have works fine and the change wouldn't gain a ton for us. 

### Details
The AwareDatetime is only used for validation. You cannot instantiate an AwareDatetime and use it like a regular `datetime` object. For example: 
`AwareDatetime(2000, 1, 1)` will result in an error:
`TypeError: AwareDatetime() takes no arguments`

However, the following works well for validation when a tz-aware datetime is required:
```python
class Model(BaseModel):
  dt: AwareDatetime
```

Since `DateTimeUTC` class has a small amount of extra functionality, `AwareDatetime` is not a drop-in replacement. For example, the timezone-conversion performed to force UTC on validation is not part of AwareDatetime. However, AwareDatetime's validation can be used to avoid performing independent validation inside DateTimeUTC. The `DateTimeUTC.construct()` method would also need to be moved out. 

One way to use AwareDatetime while preserving the timezone conversion would be to use it with Annotated.

Here's a mock-up, which is present in this PR:
```python
def set_tz_to_utc(date: datetime) -> datetime:
    """Force UTC timezone for date."""
    return date.astimezone(UTC) if date.tzinfo is not UTC else date

# Pydantic evaluates `BeforeValidator`s in RTL order within Annotated. The order could be swapped to use `AfterValidator`s.
# The validation performed by pydantic for `DateTimeUTC` will be as follows:
# 1. Run the datetime through AwareDatetime's validation.
# 2. If AwareDatetime validation passes, coerce to UTC.
DateTimeUTC = Annotated[
    datetime,
    BeforeValidator(set_tz_to_utc), 
    BeforeValidator(TypeAdapter(AwareDatetime).validate_python), 
]

# Construction method moved to standalone function.
# Any instance of `DateTimeUTC.construct()` would need to be replaced with `construct_datetime_utc()`.
def construct_datetime_utc(*args, **kwargs) -> DateTimeUTC:
    """Construct a datetime with UTC timezone."""
    if kwargs.get("tzinfo") is None:
        kwargs["tzinfo"] = UTC
    return DateTimeUTC(*args, **kwargs)
```

### Considerations
- Annotated types cannot be used with `isinstance()`. I.e., `isinstance(some_date, DateTimeUTC)` will produce an error with the above approach. 
- As seen with DateTimeUTC itself during the upgrade to Pydantic v2, the nature of Pydantic v2 type validation means a change could very well result in unexpected issues. Hopefully, offloading validation to AwareDatetime results in less maintenance overhead, not more.
- See [this search](https://github.com/search?q=org%3Aghga-de%20DateTimeUTC.construct&type=code) for a list of places where `DateTimeUTC.construct()` would need to be replaced with `construct_datetime_utc()`. It's not too long, thankfully.
- [Here](https://github.com/search?q=org%3Aghga-de+%22from+ghga_service_commons.utils.utc_dates+import+DateTimeUTC%22&type=code) is a list of all places where `DateTimeUTC` is currently imported.

### CODE REVIEW

**Review by KerstenBreuer — APPROVED** [review]
Thanks so much for this thorough investigation. Looks like a very good suggestion to me. However, would also be interested what Christoph thinks before merging.

File: src/ghga_service_commons/utils/utc_dates.py [hunk 1]
```diff
@@ -18,65 +18,49 @@
...
+from typing import Annotated
 
-from pydantic import GetCoreSchemaHandler
-from pydantic.type_adapter import TypeAdapter
-from pydantic_core import CoreSchema, core_schema
+from pydantic import AwareDatetime, TypeAdapter
+from pydantic.functional_validators import BeforeValidator
 
-__all__ = ["DateTimeUTC", "UTC", "assert_tz_is_utc", "now_as_utc"]
+__all__ = ["DateTimeUTC", "set_tz_to_utc", "UTC", "assert_tz_is_utc", "now_as_utc"]
```
Comment (Cito) [inline src/ghga_service_commons/utils/utc_dates.py:26]:
```suggestion
__all__ = ["DateTimeUTC", "convert_tz_to_utc", "UTC", "assert_tz_is_utc", "now_as_utc"]
```

File: src/ghga_service_commons/utils/utc_dates.py [hunk 2]
```diff
@@ -18,65 +18,49 @@
...
 
-    Raise a Runtimeerror if the default timezone is set differently.
+    Raise a RuntimeError if the default timezone is set differently.
     """
     if datetime.now().astimezone().tzinfo != UTC:
         raise RuntimeError("System must be configured to use UTC.")
 
 
+def set_tz_to_utc(date: datetime) -> datetime:
+    """Force UTC timezone for date."""
```
Comment (Cito) [inline src/ghga_service_commons/utils/utc_dates.py:41]:
```suggestion
def convert_tz_to_utc(date: datetime) -> datetime:
    """Convert the timezone of the given datetime object to UTC."""
```

Just to make clear that we do not forcefully set or replace the timezone, but convert a possible existing timezone (or the system timezone for a naive object) to UTC.

File: src/ghga_service_commons/utils/utc_dates.py [hunk 3]
```diff
@@ -18,65 +18,49 @@
...
+    return date.astimezone(UTC) if date.tzinfo is not UTC else date
+
+
+# A pydantic type for values that should have an UTC timezone.
+# This behaves exactly like the normal datetime type, but requires that the value
+# has a timezone and converts the timezone to UTC if necessary.
+# Validation and timezone conversion is only done via pydantic. Direct use of DateTimeUTC
+# is identical to normal datetime.
```
Comment (Cito) [inline src/ghga_service_commons/utils/utc_dates.py:49]:
```suggestion
# A Pydantic type for values that should have an UTC timezone.
# This behaves exactly like the normal datetime type, but requires a
# a timezone aware object which is converted to UTC if necessary.
# Validation and timezone conversion is only done via Pydantic.
# Direct use of DateTimeUTC is identical to normal datetime.
```

File: src/ghga_service_commons/utils/utc_dates.py [hunk 4]
```diff
@@ -18,65 +18,49 @@
...
+DateTimeUTC = Annotated[
+    datetime,
+    BeforeValidator(set_tz_to_utc),
```
Comment (Cito) [inline src/ghga_service_commons/utils/utc_dates.py:52]:
```suggestion
    # note that before validators run from right to left
    BeforeValidator(convert_tz_to_utc),
```

File: src/ghga_service_commons/utils/utc_dates.py [hunk 5]
```diff
@@ -18,65 +18,49 @@
...
+    BeforeValidator(TypeAdapter(AwareDatetime).validate_python),
+]
+
+
+def construct_datetime_utc(*args, **kwargs) -> DateTimeUTC:
```
Comment (Cito) [inline src/ghga_service_commons/utils/utc_dates.py:57]: Consider renaming this function to `utc_datetime`.

File: tests/unit/utils/test_utc_dates.py
Comment (Cito) [inline tests/unit/utils/test_utc_dates.py:1]: Can you also fix the doc string of "test_accept_aware_datetimes_in_utc"? The "does not" needs to be removed.

**Review by Cito — CHANGES_REQUESTED** [review]
Thanks @TheByronHimes for the explaration and the solution which I like as well. 

More lightweight and less fragile than subclassing AwareDatetime and does all we need.

I requesteded some cosmetic changes and suggest renaming `DateTimeUTC` to `UtcDatetime` (similar to `AwareDatetime`) and `construct_datetime_utc` to just `utc_datetime()` (similar to `datetime()`).

File: tests/unit/utils/test_utc_dates.py
Comment (TheByronHimes) [reply tests/unit/utils/test_utc_dates.py:1]: Done

File: src/ghga_service_commons/utils/utc_dates.py (hunk 5 above)
Comment (TheByronHimes) [reply src/ghga_service_commons/utils/utc_dates.py]: Done [e26ddfb](https://github.com/ghga-de/ghga-service-commons/pull/26/commits/e26ddfbacfdc24f2774fb542bfd317683c581c95)

File: src/ghga_service_commons/utils/utc_dates.py (hunk 4 above)
Comment (TheByronHimes) [reply src/ghga_service_commons/utils/utc_dates.py]: Done in [e26ddfb](https://github.com/ghga-de/ghga-service-commons/pull/26/commits/e26ddfbacfdc24f2774fb542bfd317683c581c95)

File: src/ghga_service_commons/utils/utc_dates.py (hunk 3 above)
Comment (TheByronHimes) [reply src/ghga_service_commons/utils/utc_dates.py]: Thanks, I put that in [e26ddfb](https://github.com/ghga-de/ghga-service-commons/pull/26/commits/e26ddfbacfdc24f2774fb542bfd317683c581c95)

File: src/ghga_service_commons/utils/utc_dates.py (hunk 2 above)
Comment (TheByronHimes) [reply src/ghga_service_commons/utils/utc_dates.py]: Agreed, that is better. It's in [e26ddfb](https://github.com/ghga-de/ghga-service-commons/pull/26/commits/e26ddfbacfdc24f2774fb542bfd317683c581c95)

**Review by Cito — APPROVED** [review]
Thanks a lot.

And you're right, according to PEP 8 it should be UTCDatetime, not UtcDatetime.
